src/main/scala/abalone/variant/Variant.scala

```diff
@@ -102,17 +102,88 @@ abstract class Variant private[variant] (
         )
     }.toMap
 
-  def validSideMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
+  def validMovesOf2And3(situation: Situation): Map[Pos, List[(String, Move)]] = {
+    def generateMove(orig: Pos, dest: Pos, category: String): Some[(String, Move)] =
+      Some( (category, Move(Piece(situation.player, Role.defaultRole), orig, dest, situation, boardAfter(situation, orig, dest), true, if (category == "pushout") Some(dest) else None)) )
+
+    def generateSideMoves(lineOfMarbles: List[Pos], direction: Option[String]): List[(String, Move)] = {
+      // "left" is related to the direction
+      def canLeftSideMove(pos: Pos): Boolean =
+        pos.sideMovesDirsFromDir(direction)._1.fold(false)(p => situation.board.isEmptySquare(Some(p)))
+      def canRightSideMove(pos: Pos): Boolean =
+        pos.sideMovesDirsFromDir(direction)._2.fold(false)(p => situation.board.isEmptySquare(Some(p)))
+
+      List(
+        if (lineOfMarbles.size == 3) generateSideMoves(lineOfMarbles.dropRight(1), direction)
+        else None,
+        if (!lineOfMarbles.map(canLeftSideMove).contains(false))
+          generateMove(lineOfMarbles(0), lineOfMarbles.last.sideMovesDirsFromDir(direction)._1.get, "side")
+        else None,
+        if (!lineOfMarbles.map(canRightSideMove).contains(false))
+          generateMove(lineOfMarbles(0), lineOfMarbles.last.sideMovesDirsFromDir(direction)._2.get, "side")
+        else None,
+      ).flatten
+    }
 
-  def validLineMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
-    
+    def generateMovesForNeighbours(pos: Pos, neighbour: Pos, direction: Option[String]): List[(String, Move)] = {
+      val activePlayerPieces = situation.board.piecesOf(situation.player)
+      val opponentPieces = situation.board.piecesOf(!situation.player)
+      val moves = List(
+        neighbour.dir(direction).toList.flatMap {
```

Comment:
this feels very convoluted, a different approach could be to collect the next 4 squares after the neighbour, which will either be empty or have a piece in them, and then do a match on each of the cases at the same level instead of a nested if/case statement. I think it would read better and be easier to understand? You dont have to change this code if it's functional but its another way to view the problem.

Reply:
oh yes indeed this would be better...

Reply:
I would love to implement it like that if we have enough time : let's see how Abalone progress overall in the next weeks and/or if there is a need to speed up the algos ?